src/services/schema/update/analyzer/mongo-collections-analyzer.js

```diff
@@ -85,97 +94,235 @@ function mapCollection() {
+    return fields;
+  }
 
-      return P.each(collections, async (collection) => {
-        const collectionName = getCollectionName(collection);
+  async analyzeMongoCollections(databaseConnection) {
+    const collections = await databaseConnection.collections();
+    if (collections.length === 0) {
+      throw new EmptyDatabaseError('no collections found', {
+        orm: 'mongoose',
+        dialect: 'mongodb',
+      });
+    }
 
-        // Ignore system collections and collection without a valid name.
-        if (!collectionName || isSystemCollection(collection)) {
-          return;
+    let isMongodbInstanceSupportJs = true;
+    return P.reduce(collections, async (schema, collection) => {
+      const collectionName = this.getCollectionName(collection);
+
+      // Ignore system collections and collection without a valid name.
+      if (!collectionName || this.isSystemCollection(collection)) {
+        return schema;
+      }
+
+      let analysis;
+      if (isMongodbInstanceSupportJs) {
+        analysis = await this.analyzeMongoCollectionRemotely(databaseConnection, collectionName);
+        if (analysis === MAP_REDUCE_ERROR_STRING) {
+          isMongodbInstanceSupportJs = false;
+          this.logger.warn('The analyze is running locally instead of in the db instance because your instance does not support javascript.'
+            + ' This action can takes a bit of time because it fetches all the documents.');
         }
+      }
 
-        const analysis = await analyzeMongoCollection(databaseConnection, collectionName);
-        const references = await detectReferences(databaseConnection, analysis, collectionName);
-        applyReferences(analysis, references);
-        const hasMany = await detectHasMany(databaseConnection, analysis, collectionName);
-        applyHasMany(analysis, hasMany);
-        schema[collectionName] = {
-          fields: analysis,
-          references: [],
-          primaryKeys: ['_id'],
-          options: {
-            timestamps: isUnderscored(analysis),
-          },
-        };
-      });
-    })
-    .then(() => schema);
+      if (!isMongodbInstanceSupportJs) {
+        analysis = await this.analyzeMongoCollectionLocally(databaseConnection, collectionName);
+      }
+
+      analysis = await this.applyRelationships(databaseConnection, analysis, collectionName);
+      schema[collectionName] = this.buildSchema(analysis);
+      return schema;
+    }, {});
+  }
+}
+
+async function analyzeMongoCollections(databaseConnection) {
+  const context = {
+    ...Context.inject(),
+    detectReferences,
+    applyReferences,
+    detectHasMany,
+    applyHasMany,
+    isUnderscored,
+    getMongooseTypeFromValue,
+    isOfMongooseType,
+    getMongooseArraySchema,
+    getMongooseEmbeddedSchema,
+    getMongooseSchema,
+    haveSameEmbeddedType,
+    hasEmbeddedTypes,
+    mergeAnalyzedSchemas,
+    isSystemCollection,
+    getCollectionName,
+    mapCollection,
+    reduceCollection,
+  };
```

Comment:
This sounds like an hack, there should be only one sealed context. Is there any reasons for example mongo side script?

Reply:
all functions `detectReferences`, `applyReferences` etc. should be added to the context. Then  `MongoCollectionsAnalyzer` should be added too to the context.

Reply:
✅